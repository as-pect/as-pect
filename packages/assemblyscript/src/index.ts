export type ASC = typeof import("assemblyscript/dist/asc.js") ;
export type AS = typeof import("assemblyscript/dist/assemblyscript.js");
export type TRANSFORM = typeof import("assemblyscript/dist/transform.js");

// @ts-ignore: Adding an aspect global
if (!globalThis.__aspect) globalThis.__aspect = {};

// We need to garuntee we use only a single version of the compiler

// @ts-ignore: Adding an aspect global for asc
if (!globalThis.__aspect.asc) globalThis.__aspect.asc = await import("assemblyscript/dist/asc.js");
// @ts-ignore: Adding an aspect global for asc
export const asc: ASC = globalThis.__aspect.asc;

// @ts-ignore: Adding an aspect global for asc
if (!globalThis.__aspect.assemblyscript) globalThis.__aspect.assemblyscript = await import("assemblyscript/dist/assemblyscript.js");
// @ts-ignore: Adding an aspect global for asc
export const assemblyscript: AS = globalThis.__aspect.assemblyscript;

// @ts-ignore: Adding an aspect global for asc
if (!globalThis.__aspect.transform) globalThis.__aspect.transform = await import("assemblyscript/dist/transform.js");
// @ts-ignore: Adding an aspect global for asc
export const transform: TRANSFORM = globalThis.__aspect.transform;

