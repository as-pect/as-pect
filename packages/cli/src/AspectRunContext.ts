import { IAspectConfiguration } from "./IAspectConfiguration";
import { IProcessConfiguration } from "./IProcessConfiguration";
import { IWritable } from "@as-pect/core/src/util/IWritable";

export interface IASCProps {
  stdout: IWritable;
  stderr: IWritable;
  listFiles(dirname: string, baseDir: string): string[];
  readFile(filename: string, baseDir: string): Uint8Array | null;
  writeFile(name: string, contents: Uint8Array, baseDir?: string): void;
}

export interface IASC {
  ready(): Promise<void>;
  main(
    args: string[],
    config: IASCProps,
    callback: (error: Error) => number,
  ): void;
}


export interface IOptionalRunContextProps {
  /** The assemblyscript help text generation function. */
  help: typeof import("assemblyscript/cli/util/options").help;
  /** The assemblyscript cli parse function. */
  parse: typeof import("assemblyscript/cli/util/options").parse;
  /** The result of calling the assemblyscript cli parse function. */
  parseResult: import("assemblyscript/cli/util/options").Result;
  /** The retreived as-pect config file. */
  config: Partial<IAspectConfiguration>;
  /** The assemblyscript compiler cli interface itself. */
  asc: IASC;
}

export interface IRequiredRunContextProps {
  /** Raw argv from the node process. */
  argv: string[];
  /** A set of file promises that must be resolved before the application exits. */
  files: Promise<void>[];
  /** A stripped down version of a process that contains an exit function, stdout and stderr. */
  process: IProcessConfiguration;
  /** The exit code used to exit the program. It starts as 0, and can only be set to 1 otherwise. */
  exitCode: number;
}

export type AspectRunContext = Partial<IOptionalRunContextProps> &
  IRequiredRunContextProps;
