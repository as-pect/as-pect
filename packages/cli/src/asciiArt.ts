import chalk from "chalk-template";

/**
 * @ignore
 *
 * This method prints the ascii art.
 * @param {string} version - The cli version
 */
export function printAsciiArt(): void {
  console.log(chalk`{bold.bgWhite.black ${""}       ___   _____                       __    
      /   | / ___/      ____  ___  _____/ /_   
     / /| | \\__ \\______/ __ \\/ _ \\/ ___/ __/   
    / ___ |___/ /_____/ /_/ /  __/ /__/ /_     
   /_/  |_/____/     / .___/\\___/\\___/\\__/     
                    /_/                        }
`);
}
