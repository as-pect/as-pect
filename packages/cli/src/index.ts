import { parse } from "configinator";
import { cliConfig, resolveOptionByName } from "./util/configuration";
import chalk from "chalk";
const pkg = require("../package.json");

export function asp(args: string[], readFileSync: (file: string, basename: string) => string | null): void {
  const configState = parse(args, cliConfig, {
    cwd: process.cwd(),
    readFileSync,
  });

  if (configState.diagnostics.length > 0) {
    for (const diag of configState.diagnostics) {
      console.log(chalk`{red [CLI Error]}: ${diag}`);
    }
    process.exit(1);
  }

  if (resolveOptionByName(configState, "version")) {
    console.log(chalk`{yellow [Version]}: ${pkg.version}`);
    process.exit(0);
  }

  if (!resolveOptionByName(configState, "nologo")) {
    const printAsciiArt = require("./util/asciiArt").printAsciiArt;
    printAsciiArt(pkg.version);
  }

  if (resolveOptionByName(configState, "types")) {
    const types = require("./types").types;
    types();
    process.exit(0);
  }

  if (resolveOptionByName(configState, "portable")) {
    const portable = require("./portable").portable;
    portable();
    process.exit(0);
  }

  process.exit(0);
}

if (typeof require != "undefined" && require.main == module) {
  const fs = require("fs");
  const path = require("path");

  asp(process.argv.slice(2), (file: string, basename: string): string | null => {
    const thePath = path.normalize(path.join(basename, file));
    try {
      return fs.readFileSync(thePath, "utf8");
    } catch (ex) {
      return null;
    }
  });
}