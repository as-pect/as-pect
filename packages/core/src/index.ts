export * from "./reporter/CombinationReporter";
export * from "./reporter/EmptyReporter";
export * from "./reporter/IReporter";
export * from "./reporter/SummaryReporter";
export * from "./reporter/VerboseReporter";
export * from "./test/IWarning";
export * from "./test/TestContext";
export * from "./test/TestNode";
export * from "./util/IAspectExports";
export * from "./util/IPerformanceConfiguration";
export * from "./util/ReflectedValue";
