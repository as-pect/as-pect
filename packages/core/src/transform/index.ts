// import { Transform } from "assemblyscript/cli/transform";
import {
  ClassDeclaration,
  NamespaceDeclaration,
  NodeKind,
  Parser,
  Statement,
  Transform,
  Source,
} from "./assemblyscript";
import { createStrictEqualsMember } from "./createStrictEqualsMember";
import { createAddReflectedValueKeyValuePairsMember } from "./createAddReflectedValueKeyValuePairsMember";
export = class AspectTransform extends Transform {
  /**
   * This method results in a pure AST transform that inserts a strictEquals member
   * into each ClassDeclaration.
   *
   * @param {Parser} parser - The AssemblyScript parser.
   */
  afterParse(parser: Parser): void {
    // For backwards compatibility
    let sources: Source[] = (parser as any).program
      ? (parser as any).program.sources
      : parser.sources;
    // for each program source
    for (const source of sources) {
      traverseStatements(source.statements);
    }
  }
};

function traverseStatements(statements: Statement[]): void {
  // for each statement in the source
  for (const statement of statements) {
    // find each class declaration
    if (statement.kind === NodeKind.CLASSDECLARATION) {
      // cast and create a strictEquals function
      const classDeclaration = <ClassDeclaration>statement;
      classDeclaration.members.push(createStrictEqualsMember(classDeclaration));
      classDeclaration.members.push(
        createAddReflectedValueKeyValuePairsMember(classDeclaration),
      );
    } else if (statement.kind === NodeKind.NAMESPACEDECLARATION) {
      const namespaceDeclaration = <NamespaceDeclaration>statement;
      traverseStatements(namespaceDeclaration.members);
    }
  }
}
