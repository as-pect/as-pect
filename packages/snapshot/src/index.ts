export * from "./parser";
export * from "./test/Snapshot";
export * from "./test/SnapshotDiff";
export * from "./test/unparse";
