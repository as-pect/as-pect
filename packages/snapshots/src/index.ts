export * from "./Snapshot";
export * from "./SnapshotDiff";
export * from "./SnapshotDiffResult";
