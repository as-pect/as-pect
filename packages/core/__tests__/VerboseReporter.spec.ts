import { VerboseReporter, TestContext, TestNode } from "../src";
import { createReporterModule } from "./setup/createReporterModule";
import strip from "strip-ansi";

const writer = {
  result: "",
  write(input: string): void {
    writer.result += input;
  },
  reset() {
    writer.result = "";
  },
};

class ReporterWrapper extends VerboseReporter {
  constructor() {
    super();
    this.stdout = writer;
    this.stderr = writer;
  }
  onGroupStart(group: TestNode): void {
    writer.reset();
    super.onGroupStart(group);
    const result = strip(writer.result);
    test("onGroupStart", () => expect(result).toMatchSnapshot(group.name));
    writer.reset();
  }
  onGroupEnd(group: TestNode): void {
    writer.reset();
    super.onGroupFinish(group);
    const result = strip(writer.result);
    test("onGroupEnd", () => expect(result).toMatchSnapshot(group.name));
    writer.reset();
  }
  onTestStart(group: TestNode, testResult: TestNode): void {
    writer.reset();
    super.onTestStart(group, testResult);
    const result = strip(writer.result);
    test("onTestStart", () =>
      expect(result).toMatchSnapshot(`${group.name} ${testResult.name}`));
    writer.reset();
  }
  onTestFinish(group: TestNode, testResult: TestNode): void {
    writer.reset();
    super.onTestFinish(group, testResult);
    const result = strip(writer.result);
    test("onTestFinish", () =>
      expect(result).toMatchSnapshot(`${group.name} ${testResult.name}`));
    writer.reset();
  }
  onFinish(ctx: TestContext): void {
    writer.reset();
    ctx.rootNode.visit(e => {
      e.start = 0;
      e.end = 0;
    });
    super.onFinish(ctx);
    const result = strip(writer.result);
    test("onFinish", () => expect(result).toMatchSnapshot(`finish`));
    writer.reset();
  }
  onTodo(group: TestNode, todo: string): void {
    writer.reset();
    super.onTodo(group, todo);
    const result = strip(writer.result);
    test("todo", () => expect(result).toMatchSnapshot(`${group.name}`));
    writer.reset();
  }
}

let start = new Promise<void>((resolve, reject) => {
  createReporterModule(
    "./assembly/jest-reporter.ts",
    {},
    (err, _result) => {
      if (err) {
        console.log(err);
        reject(err);
      } else {
        resolve();
      }
    },
    new ReporterWrapper(),
  );
})
  .then(
    () =>
      new Promise<void>((resolve, reject) => {
        createReporterModule(
          "./assembly/jest-reporter2.ts",
          {},
          (err, _result) => {
            if (err) {
              console.log(err);
              reject(err);
            } else {
              resolve();
            }
          },
          new ReporterWrapper(),
        );
      }),
  )
  .then(
    () =>
      new Promise<void>((resolve, reject) => {
        createReporterModule(
          "./assembly/jest-reporter3.ts",
          {},
          (err, _result) => {
            if (err) {
              console.log(err);
              reject(err);
            } else {
              resolve();
            }
          },
          new ReporterWrapper(),
        );
      }),
  );

beforeAll(() => start);
