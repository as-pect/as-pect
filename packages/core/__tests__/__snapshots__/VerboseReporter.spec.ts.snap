// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`onGroupStart 1`] = `""`;

exports[`onGroupStart: examples 1`] = `
"[Describe]: examples

     [Log]: \\"testing in a group\\"
   [Stack]: at ../assembly/assembly/internal/Reflect/Reflect.attachStackTrace (wasm-function[45]:0x11fa)
            at ../assembly/assembly/internal/log/log<~lib/string/String> (wasm-function[63]:0x16eb)
            at start:assembly/jest-reporter~anonymous|0 (wasm-function[102]:0x2ac2)
            at ../assembly/assembly/internal/Describe/describe (wasm-function[104]:0x2adb)
            at start:assembly/jest-reporter (wasm-function[105]:0x2aec)
            at ~start (wasm-function[111]:0x2b25)
"
`;

exports[`onStart: start 1`] = `""`;

exports[`onTestFinish: examples example 1 1`] = `
" [Success]: ✔ example 1
"
`;

exports[`onTestFinish: examples log value 1`] = `
" [Success]: ✔ log value
     [Log]: \\"testing in a test\\"
   [Stack]: at ../assembly/assembly/internal/Reflect/Reflect.attachStackTrace (wasm-function[45]:0x11fa)
            at ../assembly/assembly/internal/log/log<~lib/string/String> (wasm-function[63]:0x16eb)
            at start:assembly/jest-reporter~anonymous|0~anonymous|1 (wasm-function[64]:0x16fb)
            at ../assembly/assembly/internal/call/__call (wasm-function[107]:0x2afe)
"
`;

exports[`onTestFinish: examples push an error to the error array 1`] = `
" [Success]: ✔ push an error to the error array
"
`;

exports[`onTestFinish: examples something fails 1`] = `
"    [Fail]: ✖ something fails
  [Actual]: -1
[Expected]: 42
 [Message]: Here is an example failure message.
   [Stack]: RuntimeError: unreachable
            at ../assembly/assembly/internal/assert/assert (wasm-function[48]:0x1261)
            at ../assembly/assembly/internal/Expectation/Expectation<i32>#toBe (wasm-function[51]:0x12a7)
            at start:assembly/jest-reporter~anonymous|0~anonymous|3 (wasm-function[67]:0x171b)
            at ../assembly/assembly/internal/call/__call (wasm-function[107]:0x2afe)
"
`;

exports[`onTestFinish: examples something negated fails 1`] = `
"    [Fail]: ✖ something negated fails
  [Actual]: 42
[Expected]: Not 42
 [Message]: 42 is 42
   [Stack]: RuntimeError: unreachable
            at ../assembly/assembly/internal/assert/assert (wasm-function[48]:0x1261)
            at ../assembly/assembly/internal/Expectation/Expectation<i32>#toBe (wasm-function[51]:0x12a7)
            at start:assembly/jest-reporter~anonymous|0~anonymous|6 (wasm-function[100]:0x2a5a)
            at ../assembly/assembly/internal/call/__call (wasm-function[107]:0x2afe)
"
`;

exports[`onTestFinish: examples this should throw but it does not 1`] = `
"    [Fail]: ✖ this should throw but it does not
 [Message]: it should throw
"
`;

exports[`onTestFinish: examples when two things should equal each other 1`] = `
"  [Throws]: ✔ when two things should equal each other RTrace: +9
"
`;

exports[`onTestStart: examples example 1 1`] = `""`;

exports[`onTestStart: examples log value 1`] = `""`;

exports[`onTestStart: examples push an error to the error array 1`] = `""`;

exports[`onTestStart: examples something fails 1`] = `""`;

exports[`onTestStart: examples something negated fails 1`] = `""`;

exports[`onTestStart: examples this should throw but it does not 1`] = `""`;

exports[`onTestStart: examples when two things should equal each other 1`] = `""`;
